CLDC-2310 add creation method to both log import services, lettings log field import and create sales log field import

arfacamble opened the pull request:
https://digital.dclg.gov.uk/jira/browse/CLDC-2310

- add creation method to:
  - both log import services
  - lettings log field import

- create sales log field import

DEPLOY TASK ALERT this ticket contains a rake task to run when deployed

arfacamble committed:
some minor refactoring
remove methods from child class that replicate methods on the parent class
tidy up check for nil
remove gubbins and inline method body given only used once

arfacamble committed:
update import services for lettings and sales to import creation method
write tests to cover this

arfacamble committed:
create sales log field import service and associated spec file, with methods and tests for importing the creation method of logs that have already been imported

arfacamble committed:
update lettings log field import service and related spec to allow importing creation method of logs

arfacamble commented on app/services/imports/lettings_logs_field_import_service.rb:
```diff
@@ -110,31 +135,5 @@ def update_tenant_code(xml_doc)
-      else
-        str.to_i
-      end
-    end
```
these methods exist on the parent class already

kosiakkatrina replied:
nice

arfacamble commented on app/services/imports/logs_import_service.rb:
```diff
@@ -11,11 +11,7 @@ def safe_string_as_integer(xml_doc, attribute)
-      else
-        str.to_i
-      end
+      str&.to_i
```
someone sense check me here, but if str is nil we get nil, otherwise we get str.to_i

jamesrose commented on app/services/imports/lettings_logs_field_import_service.rb:
```diff
@@ -36,6 +38,29 @@ def update_offered(xml_doc)
+          if log.creation_method == "bulk upload"
+            @logger.info "lettings log #{log.id} creation method already set to bulk upload, no need for update"
+          else
+            log.update!(creation_method: "bulk upload")
```
This should be an enum.

jamesrose commented on app/services/imports/lettings_logs_field_import_service.rb:
```diff
@@ -36,6 +38,29 @@ def update_offered(xml_doc)
+        when "Manual Entry"
+          @logger.info "lettings log with old id #{old_id} entered manually, no need for update"
+        when "Bulk Upload"
+          if log.creation_method == "bulk upload"
```
```suggestion
          if log.bulk_upload_creation_method?
```

arfacamble committed:
use the methods dynamically created by active record in all relevant places, removing obsolete methods in teh process.
various tests tweaked to suppor this change.
rake task from another ticket folded into this ticket to prevent merge conflicts

kosiakkatrina commented on app/services/imports/lettings_logs_field_import_service.rb:
```diff
@@ -36,6 +38,29 @@ def update_offered(xml_doc)
+          else
+            log.creation_method_bulk_upload!
+            @logger.info "lettings log #{log.id} creation method set to bulk upload"
+          end
```
Would there be a situation where the log already has the creation method set? Wouldn't just having the else part of the statement achieve the same result?

arfacamble replied:
the creation method is always set, with a default of "single log"
49 - if the old log is manual entry, nothing to do
52 - if it's already set to bulk upload, we've been here before
54 - make it bulk upload

arfacamble replied:
oh I think I see what you mean, if the old log is bulk upload, why not just set it to bulk upload in any case and have fewer lines?

for that I was following the model of the existing methods in the field import services, which seemed to check whether the field was already set correctly and log that if so

kosiakkatrina commented on app/services/imports/lettings_logs_field_import_service.rb:
```diff
@@ -36,6 +38,29 @@ def update_offered(xml_doc)
+        end
+      else
+        @logger.warn "lettings log with old id #{old_id} not found"
+      end
```
Would a guard clause at the top instead of this else be more readable?

arfacamble replied:
I considered that and then followed the model of the other methods in these classes. I kind fo think they're equal, but it does reduce one level of nesting

arfacamble replied:
I was going to change it and then I realised that because we need to log a warning and return, you can't one line it, so you don't really reduce nesting without adding more lines, so I'll eave it as is unless you feel strongly

kosiakkatrina replied:
Would `return @logger.warn "lettings log with old id #{old_id} not found" unless log` not work?

arfacamble replied:
👍

kosiakkatrina commented on app/services/imports/logs_import_service.rb:
```diff
@@ -29,6 +25,16 @@ def compose_date(xml_doc, day_str, month_str, year_str)
+    def get_creation_method(xml_doc)
+      upload_method = meta_field_value(xml_doc, "upload-method")
+      CREATION_METHODS[upload_method]
+    end
```
`get_` is quite java like, maybe we could rename the method to just `creation_method` ?

arfacamble committed:
rename method for ruby conventions

arfacamble committed:
update PR for altered spec
upload id now decided to be a better indicator of bulk upload status, import service amended accordingly
tests updated in line with this

arfacamble committed:
update field import services in line with import services to use upload id rather than upload method as the source of truth for how a log was created

arfacamble committed:
slight refactor to reduce nesting and dodge linter complaints

arfacamble committed:
minor amendment to log creator spec in bulk upload to use enum dynamic methods
